### PR TITLE
Intro screen

### PR DESCRIPTION
**Pull request by jakearchibald**
ffs I keep forgetting GitHub auto closes PRs when the base branch is deleted. Old PR is https://github.com/GoogleChromeLabs/squoosh/pull/188.

Styled up the intro screen. Includes demo images & loading spinner component.

### CODE REVIEW

**Comment by jakearchibald** [comment]
@surma 

> Also, running this branch I don’t actually see the spinner anywhere.

It's there for the duration of the demo image loading. https://github.com/GoogleChromeLabs/squoosh/pull/191/files#diff-7983b39e7d31cf20f6fa9924363c38b4R71.

You'll see it on https://deploy-preview-191--squoosh.netlify.com/, especially if you throttle connectivity.

File: src/components/custom-els/LoadingSpinner/index.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  private _delayTimeout: number = 0;
+
+  constructor() {
+    super();
+
+    // Ideally we'd use shadow DOM here, but we're targeting browsers without shadow DOM support.
+    // You can't set attributes/content in a custom element constructor, so I'm waiting a microtask.
+    Promise.resolve().then(() => {
+      this.style.display = 'none';
+      this.innerHTML = '' +
```
Comment (developit) [inline src/components/custom-els/LoadingSpinner/index.ts]: Think it'd be worth going with something that avoids the HTML parse here?
I'm thinking something sortof like a hyperscript helper:

```js
function h(tag, className, children) {
  const el = document.createElement(tag);
  el.className = className;
  for (const child of children) el.appendChild(child);
  return el;
}

this.appendChild(
  h('div', styles.spinnerContainer, [
    h('div', styles.spinnerLayer, [
      h('div', styles.spinnerCircleClipper + ' ' + styles.spinnerLeft, [
        h('div', styles.spinnerCircle)
      ]),
      h('div', styles.spinnerGapPatch, [
        h('div', styles.spinnerCircle)
      ]),
      h('div', styles.spinnerCircleClipper + ' ' + styles.spinnerRight, [
        h('div', styles.spinnerCircle)
      ])
    ])
  ])
);
```

File: src/missing-types.d.ts [hunk 2]
```diff
@@ -20,3 +18,24 @@ declare namespace JSX {
 interface CanvasRenderingContext2D {
   filter: string;
 }
+
+// Handling file-loader imports:
+declare module '*.png' {
```
[the remark this answers is not in the document]
Comment (jakearchibald) [reply src/missing-types.d.ts]: How does file name hashing work in a `CopyWebpackPlugin` world?

**Comment by jakearchibald** [comment]
@developit ughhhhh I accidentally pushed some of my lazy-load-the-app commits to this branch. I've removed them. Sorry if it made reviewing confusing.

File: src/lib/initial-util.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// This file contains the utils that are needed for the very first rendering of the page. They're
```
[the remark this answers is not in the document]
Comment (jakearchibald) [reply src/lib/initial-util.ts]: Yeah I'm really curious about this.

File: src/components/intro/style.scss [hunk 4]
```diff
@@ -1,22 +1,193 @@
...
+  margin-bottom: 0;
+
+  @media (min-width: 460px) {
+    font-size: 50.6px;
+    padding: 0 40px;
+  }
+}
+
+.select-button {
+  all: unset;
```
[the remark this answers is not in the document]
Comment (jakearchibald) [reply src/components/intro/style.scss]: Oh bollocks, I didn't realise Edge didn't support this. Tbh I ran into a Safari bug here too https://bugs.webkit.org/show_bug.cgi?id=158782.

I'll do something else.

File: src/components/intro/index.tsx [hunk 5]
```diff
@@ -1,26 +1,134 @@
 import { h, Component } from 'preact';
+
+import { bind, linkRef, Fileish } from '../../lib/initial-util';
+import '../custom-els/LoadingSpinner';
+
+import logo from './imgs/logo.svg';
```
[the remark this answers is not in the document]
Comment (jakearchibald) [reply src/components/intro/index.tsx]: I think the hashing is pretty important here. These resources really benefit from caching.

File: src/components/custom-els/LoadingSpinner/index.ts (hunk 1 above)
Comment (jakearchibald) [reply src/components/custom-els/LoadingSpinner/index.ts]: https://jsperf.com/innerhtml-vs-createelement-test/30 - parsing is really fast, so I'm happy to go with readability here.

Comment (developit) [reply src/components/custom-els/LoadingSpinner/index.ts]: hmm - not sure I actually trust that benchmark. It doesn't seem to hold up:

https://esbench.com/bench/59b3233399634800a0349038
image

Comment (developit) [reply src/components/custom-els/LoadingSpinner/index.ts]: agreed on the readability bit though, and I don't think the performance difference ends up being super meaningful either way here given that it only happens on creation.

**Comment by developit** [comment]
Nah, no prob, I just review commit by commit.

**Pull request merged**